drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,528 @@
+#include <logging/log.h>
+
+#ifdef CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+#include <irq.h>
+#endif
+
+#include "hx711.h"
+
+LOG_MODULE_REGISTER(HX711, CONFIG_LOG_DEFAULT_LEVEL);
+
+static struct hx711_data hx711_data = {
+		.reading = 0,
+		.offset = CONFIG_HX711_OFFSET,
+		.slope = {
+				.val1 = CONFIG_HX711_SLOPE_INTEGER,
+				.val2 = CONFIG_HX711_SLOPE_DECIMAL,
+		},
+		.gain = CONFIG_HX711_GAIN,
+		.rate = CONFIG_HX711_SAMPLING_RATE
+};
+
+static const struct hx711_config hx711_config = {
+		.dout_pin = DT_INST_GPIO_PIN(0, dout_gpios),
+		.dout_ctrl = DT_INST_GPIO_LABEL(0, dout_gpios),
+		.dout_flags = DT_INST_GPIO_FLAGS(0, dout_gpios),
+
+		.sck_pin = DT_INST_GPIO_PIN(0, sck_gpios),
+		.sck_ctrl = DT_INST_GPIO_LABEL(0, sck_gpios),
+		.sck_flags = DT_INST_GPIO_FLAGS(0, sck_gpios),
+
+#if DT_NODE_HAS_PROP(HX711_NODE, rate_gpios)
+		.rate_pin = DT_INST_GPIO_PIN(0, rate_gpios),
+		.rate_ctrl = DT_INST_GPIO_LABEL(0, rate_gpios),
+		.rate_flags = DT_INST_GPIO_FLAGS(0, rate_gpios),
+#endif
+};
+
+/**
+ * @brief Check if HX711 data is ready
+ *
+ * @param data Pointer to the hx711 driver data structure
+ *
+ * @retval true when data is available,
+ * @retval false if data is unavailable
+ */
+
+static bool hx711_is_ready(struct hx711_data *data)
+{
+	return !gpio_pin_get(data->dout_gpio, hx711_config.dout_pin);
+}
+
+/**
+ * @brief Wait until HX711 data is ready
+ *
+ * @param data Pointer to the hx711 driver data structure
+ * @param usec Usecs to wait before returning.
+ *
+ * @retval 0 When data is available,
+ * @retval -EACCES error if usec's are elapsed with no data available.
+ */
+static int hx711_wait_until_ready(struct hx711_data *data, uint32_t usec)
+{
+	while (!hx711_is_ready(data) && usec >= 0) {
+		k_busy_wait(1);
+		usec--;
+	}
+	;
+	if (usec == 0) {
+		return -EACCES;
+	} else   {
+		return 0;
+	}
+}
+
+/**
+ * @brief Send a pulse on the SCK pin.
+ *
+ * @param data Pointer to the hx711 driver data structure
+ *
+ * @retval The value of the DOUT pin (HIGH or LOW)
+ *
+ */
+static int hx711_cycle(struct hx711_data *data)
+{
+	/* SCK set HIGH */
+	gpio_pin_set(data->sck_gpio, hx711_config.sck_pin, true);
+	k_busy_wait(1);
+
+	/* SCK set LOW */
+	gpio_pin_set(data->sck_gpio, hx711_config.sck_pin, false);
+	k_busy_wait(1);
+
+	/* Return DOUT pin state */
+	return gpio_pin_get(data->dout_gpio, hx711_config.dout_pin);
+}
+
+
+/**
+ * @brief Read HX711 data. Also sets GAIN for the next cycle.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to fetch data for.
+ *             Only SENSOR_CHANNEL_WEIGHT is available.
+ *
+ * @retval 0 on success,
+ * @retval -EACCES error if 100 usec's are elapsed with no data available.
+ *
+ */
+static int hx711_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	int ret = 0;
+	u32_t count = 0;
+	int i;
+
+
+	struct hx711_data *data = dev->driver_data;
+
+	ret = hx711_wait_until_ready(data, 100);
+
+
+	/* Clock data out. Optionally disable interrupts */
```

Comment:
why optionally disable interrupts?

Reply:
If the **SCK** pin stays **HIGH** for longer than **50 µs** then the **HX711** switches off.

I introduced **CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING** to let the user decide if measuring weight is more important than other aspects of their software that might be IRQ based.

I don't really expect issues but to make sure I disable IRQ for the solution I am working on.

Someone else might not want this, so he can let IRQ's on.
(default **DISABLE_INTERRUPTS_WHILE_POLLING** is **n**)

Leaving as is for now.